substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JDKContainerSubstitutions.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2022, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jdk;
+
+import org.graalvm.nativeimage.Platform.LINUX;
+import org.graalvm.nativeimage.Platforms;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.RecomputeFieldValue.Kind;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(className = "jdk.internal.platform.cgroupv1.CgroupV1Subsystem")
+@Platforms(LINUX.class)
+final class Target_jdk_internal_platform_cgroupv1_CgroupV1Subsystem {
+    @Alias //
+    @RecomputeFieldValue(kind = Kind.Reset) //
+    private static volatile Target_jdk_internal_platform_cgroupv1_CgroupV1Subsystem INSTANCE;
+}
+
+@TargetClass(className = "jdk.internal.platform.cgroupv2.CgroupV2Subsystem")
+@Platforms(LINUX.class)
+final class Target_jdk_internal_platform_cgroupv2_CgroupV2Subsystem {
+    @Alias //
+    @RecomputeFieldValue(kind = Kind.Reset) //
+    private static volatile Target_jdk_internal_platform_cgroupv2_CgroupV2Subsystem INSTANCE;
+}
+
+@TargetClass(className = "jdk.jfr.internal.instrument.JDKEvents")
+@Platforms(LINUX.class)
+final class Target_jdk_jfr_internal_instrument_JDKEvents {
+    @Alias //
+    @RecomputeFieldValue(kind = Kind.Reset) //
+    private static Target_jdk_internal_platform_Metrics containerMetrics;
```

Comment:
How is this re-initialized at run time? Unless I'm missing something, this will just disable container events.

Reply:
It's initialized with `PlatformRecorder()` => `JDKEvents.initialize()` => `initializeContainerEvents()`. What this change does is preventing build-time state to populate to runtime unless I'm missing something. In JVM mode it's set to `null` here:
https://github.com/openjdk/jdk17u-dev/blob/master/src/jdk.jfr/share/classes/jdk/jfr/internal/instrument/JDKEvents.java#L116

https://github.com/oracle/graal/issues/5396 has a reproducer and according to my testing it's fixed with this. Have you determined otherwise?

Reply:
And `PlatformRecorder` is initialized at run-time?

Reply:
That's my understanding, yes. When I run the reproducer:

```
[root@7de05057c67d /]# /opt/target/debug-symbols-smoke -XX:+FlightRecorder -XX:StartFlightRecording=filename=flight-native.jfr -XX:FlightRecorderLogging=jfr
[info][jfr] Added periodic hook for EveryChunkPeriodEvents(12336)
[info][jfr] Added periodic hook for EndChunkPeriodEvents(12314)
[info][jfr] Flight Recorder initialized
[info][jfr] Repository base directory: /tmp
[info][jfr] Started recording "1" (1) {dumponexit=true, filename=/flight-native.jfr}
Q to quit
```

Note the line `Flight Recorder initialized` which comes from https://github.com/openjdk/jdk17u-dev/blob/master/src/jdk.jfr/share/classes/jdk/jfr/FlightRecorder.java#L188 at which point a new instance of `PlatformRecorder()` would already been created, triggering the initialization chain.

Reply:
Thanks for checking!

Reply:
Note that this initially came from https://github.com/oracle/graal/pull/5400#discussion_r1018769862 where I've opted for class re-init of the cgroup instances (then changed it to instance reset). However, I'll run some more tests with JFR enabled at image build time. It might be we need to also reset https://github.com/openjdk/jdk17u-dev/blob/master/src/jdk.jfr/share/classes/jdk/jfr/internal/instrument/JDKEvents.java#L117

Reply:
> I'll run some more tests with JFR enabled at image build time.

This also works with an image produced with JFR enabled at image build time according to my testing. So I think this is resolved?

Reply:
Again, thanks for checking. Yes, I think this is resolved. I've put the PR in the merge queue. Enjoy your weekend. :)

Reply:
I had to take it out of the merge queue and will be on PTO until the end of the month. Internal reviewers noted:

>I hope I'm not missing something obvious, but based on the `JDKEvents.initialize()` source code, this substitution just doesn't make sense - it's either redundant or incomplete. ...

and

>... I checked both the JDK 17 and JDK 19 sources and as you stated, this does not make any sense. There are two cases:
a. Native image build with JFR at runtime only: `JDKEvents.initialize()` is not executed at image build time, so the `containerMetrics` are also not initialized. Resetting containerMetrics is useless but not harmful.
b. Native image build with JFR at both build and runtime: `JDKEvents.initialize()` is executed at image build time and `containerMetrics` is initialized. Resetting `containerMetrics` alone is not sufficient, and reexecuting `JDKEvents.initialize()` can be harmful.

Reply:
Thanks. Apparently in my testing I've used an old image :-( Sorry about that. Yes, those JDKEvents resets are necessary to cater the case where JFR is enabled at build and runtime. As expected, https://github.com/openjdk/jdk17u-dev/blob/master/src/jdk.jfr/share/classes/jdk/jfr/internal/instrument/JDKEvents.java#L117 needed resets too as well as the list recording registered hooks. The updated patch should have this now.

Reply:
If `JDKEvents.initialize()` was executed at build time and is then reexecuted at run-time, I think that this can cause issues. Doesn't for example the following call chain throw an `IllegalArgumentException("Hook has already been added")` on JDK 19?
`JDKEvents.initialize()` -> `JDKEvents.initializeContainerEvents()` -> `RequestEngine.addTrustedJDKHook(...)` -> `RequestEngine.addHookInternal(...)`

Reply:
By `disregard the example below` you meant this, right?

```
I think that reexecuting `JDKEvents.initialize()` at runtime can cause issues if the method was already executed at image build time. For example, doesn't the following call chain throw an `IllegalArgumentException("Hook has already been added")` on JDK 19?

`JDKEvents.initialize()` -> `RequestEngine.addTrustedJDKHook(...)` -> `RequestEngine.addHookInternal(...)`
```